src/backend/address_space_core.h

```diff
@@ -57,45 +51,77 @@ namespace snmalloc
     }
 
     /**
-     * Adds a block to `ranges`.
+     * Set next pointer for a power of two address range.
+     *
+     * This abstracts the use of either
+     * - the pagemap; or
+     * - the first pointer word of the block
+     * to store the next pointer for the list of unused address space of a
+     * particular size.
      */
-    template<SNMALLOC_CONCEPT(ConceptPAL) PAL>
-    void add_block(size_t align_bits, CapPtr<void, CBChunk> base)
+    template<typename Pagemap>
+    void set_next(
+      size_t align_bits,
+      CapPtr<void, CBChunk> base,
+      CapPtr<void, CBChunk> next,
+      Pagemap& pagemap)
     {
-      check_block(base, align_bits);
-      SNMALLOC_ASSERT(align_bits < 64);
-      if (ranges[align_bits][0] == nullptr)
+      if (align_bits >= MIN_CHUNK_BITS)
       {
-        // Prefer first slot if available.
-        ranges[align_bits][0] = base;
+        // The pagemap stores MetaEntrys, abuse the metaslab field to be the
+        // next block in the stack of blocks.
```

Comment:
```suggestion
        // next block in the stack of blocks.
        //
        // The pagemap entries here have nullptr (i.e., fake_large_remote) as
        // their remote, and so other accesses to the pagemap (by external_pointer,
        // for example) will not attempt to follow this "Metaslab" pointer.
```

Is that right?

Reply:
Yes, but also this is memory that is completely controlled by snmalloc, so I don't think anyone should query it with `external_pointer`.  It is not going to be a stack, global, mmap that is done elsewhere in the program.  Though, not entirely sure what the correct full spec of `external_pointer` is for that argument.